### PR TITLE
feat: switch to safe-stable-stringify for safer serialization

### PR DESCRIPTION
**Pull request by trentm**
This drops the schema-based fast-json-stringify usage, which could
potentially be faster, but will throw on circular references. The
safer serialization is more appropriate for logging libs. Note that
currently both pino and winston themselves use safe-stable-stringify.

Add explicit dep for triple-beam, rather than getting lucky.
Use a Format class to avoid needing a dep on winston or logform
for the 'winston.format(...)' call.

Document subtle diffs with our stringifier and winston.format.json().

Closes: https://github.com/elastic/ecs-logging-nodejs/issues/104
Closes: https://github.com/elastic/ecs-logging-nodejs/pull/144
Closes: https://github.com/elastic/ecs-logging-nodejs/pull/145
Closes: https://github.com/elastic/ecs-logging-nodejs/pull/108

# checklist

TODO:

- [x] add test cases for circular refs for each to show it works
- [x] consider using `logform.json()` instead, as suggested at https://github.com/elastic/ecs-logging-nodejs/issues/104#issuecomment-1756666501

> Related question, why doesn't the helper just use the same winston.format export to do it's work? Seems like asking for trouble.

### CODE REVIEW

**Comment by trentm** [comment]
The tests here are expected to fail for node <16, because logger packages are relying on getting not-yet-published changes in the packages/ecs-helpers package. Node >=16 have an npm which supports `npm --workspaces install` linking, so they work correctly.

As long as I publish `@elastic/ecs-helpers@2.0.0` before publishing new versions of the logger packages, then all will work fine for all supported node versions.

**Comment by trentm** [comment]
TODO: consider using `logform.json()` instead, as suggested at https://github.com/elastic/ecs-logging-nodejs/issues/104#issuecomment-1756666501

> Related question, why doesn't the helper just use the same winston.format export to do it's work? Seems like asking for trouble.

**Comment by trentm** [comment]
> [x] consider using `logform.json()` instead, as suggested at...

I considered it and I'm **not** doing so for now because (a) there are a couple diffs in the two uses of safe-stable-stringify (documented in a comment for now) and (b) it would require some heavier work to the winston formatter code than I want to do in this PR. I'll revisit this in a bit on https://github.com/elastic/ecs-logging-nodejs/issues/57

File: packages/ecs-morgan-format/package.json [hunk 1]
```diff
@@ -37,7 +37,8 @@
     "node": ">=10"
   },
   "dependencies": {
-    "@elastic/ecs-helpers": "^1.1.0"
+    "@elastic/ecs-helpers": "^2.0.0",
```
Comment (trentm) [inline packages/ecs-morgan-format/package.json:40]: Reviewer note: As mentioned at https://github.com/elastic/ecs-logging-nodejs/pull/155#issuecomment-1769521395 the tests for node <16 are expected to fail on this.

Comment (david-luna) [reply packages/ecs-morgan-format/package.json:40]: Could we do the package linking ourselves? at least for CI?

Comment (trentm) [reply packages/ecs-morgan-format/package.json:40]: I don't think it is worth the complexity, frankly. We'd have to attempt the npm install, trap that it has failed, check that the `dependencies["@elastic/ecs-helpers"]` entries in each of the logger's "package.json" files points to a version that matches the version in "packages/ecs-helpers/package.json#name", then link in that case -- reproducing what `npm --workspaces ...` does.

This is only necessary for a subset of changes when there is a required dependency between the ecs-helpers and logger packages, and only for now EOL versions of node.

A perhaps better alternative would be to install `npm@7` on those older versions in CI.  I could try that. (I'm hoping I can get this in before the weekend.)

Comment (trentm) [reply packages/ecs-morgan-format/package.json:40]: > A perhaps better alternative would be to install `npm@7` on those older versions in CI.

That worked. Commit d87337c92e3e96678ebe5668495d497699a7da4f

**Review by david-luna — APPROVED** [review]

**Comment by trentm** [comment]
`@elastic/ecs-helpers@2.0.0` published. It'll be a few days before I publish versions of the loggers that are using this as a dep.

**Pull request merged**